source/develop/release-management/features/storage/cd-block-storage.md

```diff
@@ -0,0 +1,228 @@
+        <ovirt-vm:volumeChain>
+            <ovirt-vm:volumeChainNode>
+                <ovirt-vm:domainID>88252cf6-381e-48f0-8795-a294a32c7149</ovirt-vm:domainID>
+                <ovirt-vm:imageID>89f05c7d-b961-4935-993f-514499024515</ovirt-vm:imageID>
+                <ovirt-vm:leaseOffset type="int">105906176</ovirt-vm:leaseOffset>
+                <ovirt-vm:leasePath>/dev/88252cf6-381e-48f0-8795-a294a32c7149/leases</ovirt-vm:leasePath>
+                <ovirt-vm:path>/rhev/data-center/mnt/blockSD/88252cf6-381e-48f0-8795-a294a32c7149/images/89f05c7d-b961-4935-993f-514499024515/626a493f-5214-4337-b580-96a1ce702c2a</ovirt-vm:path>
+                <ovirt-vm:volumeID>626a493f-5214-4337-b580-96a1ce702c2a</ovirt-vm:volumeID>
+            </ovirt-vm:volumeChainNode>
+        </ovirt-vm:volumeChain>
+    </ovirt-vm:device>
+```
+
+When CD is not loaded, this element is still present (as CD-ROM is always present), but is empty, e.g.:
+
+```
+    <ovirt-vm:device devtype="disk" name="sdc"/>
+```
+
+### Bugs
+
+- [Bug 1589763: Error changing CD for a running VM when ISO image is on a block domain](https://bugzilla.redhat.com/1589763)
+- [Bug 1868643: Volume not deactivated after ejecting CDROM](https://bugzilla.redhat.com/1868643)
+
+
+### Functionality added with this feature
+
+This feature adds new APIs to vdsm for loading, ejecting and changing CD.
+With these functions, all aforementioned functionality will work properly.
+
+#### Vdsm
+
+- Add VM.insertCD API
+- Add VM.ejectCD API
+- Updated recovery flow to check `in_progress` flag for CDs
+
+
+### Flows
+
+#### VM with empty tray, insert CD
+
+1. engine sends `insertCD` command, specifying ISO by PDIV format
+1. add new drivespec to VM metadata with new attribute `in_progress`
+```
+    <ovirt-vm:device devtype="disk" name="sdc" in_progress="yes">
+```
+1. prepare new drivespec
+  - if failed:
+    1. remove drivespec (log errors)
+    1. fail the flow with original error
+1. loads CD into VM
+  - if failed:
+    1. teardown the volume (log error)
+    1. remove the drive spec (log error)
+    1. fail the flow with the original error
+1. remove `in-progress` attribute from VM metadata
+  - if failed: log the error and succeed
+
+
+#### VM with loaded CD, eject CD
+
+1. engine sends `ejectCD`, specifying ISO path by PDIV format 
+1. vdsm updates VM metadata with new attribute `in_progress`:
+```
+    <ovirt-vm:device devtype="disk" name="sdc" in_progress="yes">
+```
+1. vdsm eject CD from VM (calling libvirt `updateDeviceFlags()`)
+  - if failed (e.g. timeout?):
+    1. remove the `in_progress` flag (log error)
+    1. fail with original error
+1. tear down drivespec
+  - if tearing down volume fails: log error (volume used by another VM)
+1. remove drivespec from VM metadata
+  - if failed: log error and succeed
+
+
+#### VM recovery flow
+
+Try to cleanup storage leftovers from insert and remove operations without changing VM state.
+
+1. metadata in progress, volume is not active
```

Comment:
How exactly is "active" related to the flows? Is it "not active" -> "prepare new drivespec" -> "active" -> "teardown" -> "not active"? It would be worth to clarify and/or using common terminology to avoid any confusion.

Reply:
I'm not sure I understand here. By active I mean active volume. On block devices, we need to active LVM volume before use and deactivate once not used, so this covers a situation when we load CD, add `in_progress` attribute into metadata, but vdsm fails/is killed/whatever before it activates the volume.

As for terminology, activate/deactivate volume is common terminology, unless I'm missing something or both of us speak about different things.

Reply:
In vdsm API we use prepare and teardown. The underlyng term in lvm level is activation/deactivation.
In this document we better use prepare/teardown.